src/TotalDissipationRateEquationSystem.C

```diff
@@ -0,0 +1,652 @@
+    copyStateAlg_.push_back(theCopyAlg);
+  }
+}
+
+//--------------------------------------------------------------------------
+//-------- register_interior_algorithm -------------------------------------
+//--------------------------------------------------------------------------
+void
+TotalDissipationRateEquationSystem::register_interior_algorithm(
+  stk::mesh::Part* part)
+{
+
+  // types of algorithms
+  const AlgorithmType algType = INTERIOR;
+
+  ScalarFieldType& tdrNp1 = tdr_->field_of_state(stk::mesh::StateNP1);
+  VectorFieldType& dedxNone = dedx_->field_of_state(stk::mesh::StateNone);
+
+  if (edgeNodalGradient_ && realm_.realmUsesEdges_)
+    nodalGradAlgDriver_.register_edge_algorithm<ScalarNodalGradEdgeAlg>(
+      algType, part, "tdr_nodal_grad", &tdrNp1, &dedxNone);
+  else
+    nodalGradAlgDriver_.register_elem_algorithm<ScalarNodalGradElemAlg>(
+      algType, part, "tdr_nodal_grad", &tdrNp1, &dedxNone, edgeNodalGradient_);
+
+  // solver; interior contribution (advection + diffusion)
+  if (!realm_.solutionOptions_->useConsolidatedSolverAlg_) {
+
+    std::map<AlgorithmType, SolverAlgorithm*>::iterator itsi =
+      solverAlgDriver_->solverAlgMap_.find(algType);
+    if (itsi == solverAlgDriver_->solverAlgMap_.end()) {
+      SolverAlgorithm* theAlg = NULL;
+      if (realm_.realmUsesEdges_) {
+        const bool useAvgMdot = realm_.solutionOptions_->turbulenceModel_ == SST_AMS;
+        theAlg = new ScalarEdgeSolverAlg(
+          realm_, part, this, tdr_, dedx_, evisc_, useAvgMdot);
+      } else {
+        throw std::runtime_error(
+          "TDREQS: Attempt to use non-NGP element solver algorithm");
+      }
+      solverAlgDriver_->solverAlgMap_[algType] = theAlg;
+
+      // look for fully integrated source terms
+      std::map<std::string, std::vector<std::string>>::iterator isrc =
+        realm_.solutionOptions_->elemSrcTermsMap_.find(
+          "total_dissipation_rate");
+      if (isrc != realm_.solutionOptions_->elemSrcTermsMap_.end()) {
+        throw std::runtime_error(
+          "TotalDissipationElemSrcTerms::Error can not use element source "
+          "terms for an edge-based scheme");
+      }
+    } else {
+      itsi->second->partVec_.push_back(part);
+    }
+
+    // Check if the user has requested CMM or LMM algorithms; if so, do not
+    // include Nodal Mass algorithms
+    std::vector<std::string> checkAlgNames = {
+      "total_dissipation_rate_time_derivative",
+      "lumped_total_dissipation_rate_time_derivative"};
+    bool elementMassAlg = supp_alg_is_requested(checkAlgNames);
+    auto& solverAlgMap = solverAlgDriver_->solverAlgMap_;
+    process_ngp_node_kernels(
+      solverAlgMap, realm_, part, this,
+      [&](AssembleNGPNodeSolverAlgorithm& nodeAlg) {
+        if (!elementMassAlg)
+          nodeAlg.add_kernel<ScalarMassBDFNodeKernel>(realm_.bulk_data(), tdr_);
+
+        if (KE == realm_.solutionOptions_->turbulenceModel_) {
```

Comment:
probably don't need this if/else, I guess this is how `KE_AMS` is being captured here

Reply:
Going to leave this because KE_AMS will slot in here